### PR TITLE
added basic file system implementation

### PR DESCRIPTION
**Pull request by xh3b4sd**
This is to provide file system implementations we can abstract and test against.

### CODE REVIEW

**Comment by xh3b4sd** [comment]
RFR @JosephSalisbury @zeisss

File: file-system/spec/spec.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+// implementations.
+package filesystemspec
+
+import (
+	"os"
+)
+
+// FileSystem implements file system operations.
+type FileSystem interface {
+	// ReadDir implements os.ReadDir.
```
Comment (zeisss) [inline file-system/spec/spec.go]: this godoc is misleading, as the references stdlib things are functions, not interfaces. How about "ReadDir is the equivalent to os.ReadDir"

File: file-system/fake/file.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+import (
+	"bytes"
+	"errors"
+	"os"
+	"time"
+)
+
+var (
+	ErrFileOperationNotPermitted = errors.New("fake file does not exist and therefore this operation is not permitted")
```
Comment (zeisss) [inline file-system/fake/file.go]: do we want to put these into the `error.go` file? (or even into the spec?)

Comment (zeisss) [inline file-system/fake/file.go]: If we would use `errgo.Newf()` here, we get the codeline where the error is defined too

File: file-system/fake/fake.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+	return nil, maskAny(pathErr)
+}
+
+func (f *fake) WriteFile(filename string, bytes []byte, perm os.FileMode) error {
+	dir := filepath.Dir(filename)
+	base := filepath.Base(filename)
+
+	if _, ok := f.Storages[dir]; !ok {
+		f.Storages[dir] = map[string][]byte{}
```
Comment (zeisss) [inline file-system/fake/fake.go:72]: so this has the effect that the folder of `WriteFile` is silently created, which I like. I don't think we have the same behavior in `real.WriteFile()` though.

Comment (xh3b4sd) [inline file-system/fake/fake.go:72]: There is a slight difference between the implementations yes. All I want is the abstraction and the tests. Both things are accomplished. I am fine with that.

File: file-system/spec/spec.go (hunk 1 above)
Comment (xh3b4sd) [inline file-system/spec/spec.go]: Fixed

File: file-system/fake/file.go (hunk 2 above)
Comment (xh3b4sd) [inline file-system/fake/file.go]: Copy/paste. Removed errors since they are not used.

**Comment by xh3b4sd** [comment]
I would like to merge this. Feedback please.

File: file-system/fake/fake.go (hunk 3 above)
Comment (zeisss) [inline file-system/fake/fake.go:72]: But having different behaviors in the implementation makes your tests useless if in production we use a different backend that behaves different.

File: file-system/fake/file.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+// Fake represents are readable in memory version of os.File which
+// can be used for testing.
+type file struct {
+	Name    string
+	Mode    os.FileMode
+	ModTime time.Time
+	Buffer  *bytes.Reader
+}
+
+// newFile creates a new file instances based on a file name
```
Comment (zeisss) [inline file-system/fake/file.go]: s/instances/instance/

**Comment by zeisss** [comment]
found a typo, still LGTM

File: file-system/fake/fake.go (hunk 3 above)
Comment (xh3b4sd) [inline file-system/fake/fake.go:72]: Is this really critical to you? What do you suggest to do then?

File: file-system/fake/file.go (hunk 4 above)
Comment (xh3b4sd) [inline file-system/fake/file.go]: Done

File: file-system/fake/fake.go (hunk 3 above)
Comment (zeisss) [inline file-system/fake/fake.go:72]: Not critical for now, you got a LGTM :cake: 

Still, I would prefer either a) add a comment to the Spec and just state that the behavior is optional or b) add a `os.MkdirAll()` call to the real implementation to align it.

Comment (xh3b4sd) [inline file-system/fake/fake.go:72]: Done

**Pull request merged**